core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -864,20 +870,16 @@ object IO {
    */
   final def bracket0[E, A, B](
     acquire: IO[E, A]
-  )(release: (A, Option[Either[E, B]]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] =
-    Ref[Option[(A, Option[Either[E, B]])]](None).flatMap { m =>
+  )(release: (A, ExitResult[E, B]) => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] =
+    Ref[Option[(A, ExitResult[E, B])]](None).flatMap { m =>
       (for {
-        a <- acquire
-              .flatMap(a => m.set(Some((a, None))).const(a))
-              .uninterruptibly
-        b <- use(a).attempt.flatMap(
-              eb =>
-                m.set(Some((a, Some(eb)))) *> (eb match {
-                  case Right(b) => IO.now(b)
-                  case Left(e)  => fail[E](e)
-                })
-            )
-      } yield b).ensuring(m.get.flatMap(_.fold(unit) { case (a, r) => release(a, r) }))
+        a <- acquire.uninterruptibly
```

Comment:
Damn, so you still need the `Option` around `ExitResult` 😞 _Or_, if a fiber is interrupted it will end up as `ExitResult.Terminated`. What do you think of priming the ref with a default result like this:
```
a <- acquire
  .flatMap(a => m.set(Some(a, ExitResult.Terminated(Nil))).const(a))
  .uninterruptibly
```
?

Reply:
Priming with `Terminated(Nil)` works. We will lose the exact cause of interruption (if it's specified). 

The alternative is do the fork of `use` inside the uninterruptible acquire section, and add an `onComplete` that will capture the cause of failure.

Then you have a guarantee (as per contract of `onComplete`) that whether or not `use` is interrupted, the `onComplete` handler will be invoked.

Maybe we should do this to be as correct as possible. The `bracket` variant will be faster for the common case, anyway, because it has no forks. And we may well end up specializing `bracket0` as a primitive for JS reasons.